### PR TITLE
KB-2925 | New API  for the OTP generation for email/phone verification

### PR DESCRIPTION
**Pull request by tarentomaheshvakkund**
1. Replicated the existing APIS and made the necessary code changes for contextAttributes, contextTypes and accessToken.

### CODE REVIEW

File: controller/app/controllers/otp/validator/OtpRequestValidator.java [hunk 1]
```diff
@@ -102,4 +102,39 @@ private void validateType(String type) {
...
     }
   }
+
+  public void validateGenerateOtpRequestV3(Request otpRequest) {
+    commonValidationV3(otpRequest, false);
+    validateTemplateId(otpRequest);
+  }
+
+
+  private void commonValidationV3(Request otpRequest, boolean isOtpMandatory) {
```
Comment (karthik-tarento) [inline controller/app/controllers/otp/validator/OtpRequestValidator.java]: This method suppose to be 
validateParam(
            (String) otpRequest.getRequest().get(JsonKey.CONTEXT_TYPE),
            ResponseCode.mandatoryParamsMissing,
            JsonKey.KEY);
 commonValidation(otpRequest, false);
    validateTemplateId(otpRequest);

File: core/platform-common/src/main/java/org/sunbird/operations/ActorOperations.java [hunk 2]
```diff
@@ -73,6 +73,8 @@ public enum ActorOperations {
   GENERATE_OTP("generateOTP", "OTPCRT"),
   VERIFY_OTP("verifyOTP", "OTPVERFY"),
   SEND_OTP("sendOTP", "OTPNOTI"),
+  GENERATE_OTP_V3("generateOTPV3", "OTPCRT"),
```
Comment (karthik-tarento) [inline core/platform-common/src/main/java/org/sunbird/operations/ActorOperations.java]: check the same value can be used or different one required

File: service/src/main/java/org/sunbird/actor/otp/OTPActor.java [hunk 3]
```diff
@@ -240,4 +254,185 @@ private String getType(String type) {
...
+    logger.debug(request.getRequestContext(), "OTPActor:generateOTP method call start.");
+    // Extract necessary parameters from the request.
+    String type = (String) request.getRequest().get(JsonKey.TYPE);
+    String key = (String) request.getRequest().get(JsonKey.KEY);
+    String userId = (String) request.getRequest().get(JsonKey.USER_ID);
+    String contextType = (String) request.getRequest().get(JsonKey.CONTEXT_TYPE);
+    List<String> contextAttributesList = (new ObjectMapper()).convertValue(request.get(JsonKey.CONTEXT_ATTRIBUTES),
+            new TypeReference<>() {
+            });
+    String contextAttributes = String.join(",", contextAttributesList);
```
Comment (karthik-tarento) [inline service/src/main/java/org/sunbird/actor/otp/OTPActor.java]: Null check is required

File: service/src/main/java/org/sunbird/actor/otp/OTPActor.java [hunk 4]
```diff
@@ -240,4 +254,185 @@ private String getType(String type) {
...
+    // If no details found, generate a new OTP, log its generation, insert the OTP details, and log the insertion.
+    if (MapUtils.isEmpty(details)) {
+      otp = OTPUtil.generateOTP(request.getRequestContext());
+      logger.info(
+              request.getRequestContext(),
+              "OTPActor:generateOTP: new otp generated for Key = "
+                      + OTPUtil.maskId(key, type)
+                      + " & OTP = "
+                      + OTPUtil.maskOTP(otp));
+      otpService.insertOTPDetailsV3(type, key, otp, contextType, contextAttributes, request.getRequestContext());
```
Comment (karthik-tarento) [inline service/src/main/java/org/sunbird/actor/otp/OTPActor.java:308]: Do not send contextAttributes if empty

File: service/src/main/java/org/sunbird/dao/otp/impl/OTPDaoImpl.java [hunk 5]
```diff
@@ -90,4 +90,89 @@ public void updateAttemptCount(Map<String, Object> otpDetails, RequestContext co
...
+  @Override
+  public void insertOTPDetailsV3(String type, String key, String otp, String contextType, String contextAttributes, RequestContext context) {
+    // Create a map to store OTP details.
+    Map<String, Object> request = new HashMap<>();
+    // Populate the map with OTP details.
+    request.put(JsonKey.TYPE, type);
+    request.put(JsonKey.KEY, key);
+    request.put(JsonKey.OTP, otp);
+    request.put(JsonKey.CONTEXT_TYPE, contextType);
+    request.put(JsonKey.CONTEXT_ATTRIBUTES, contextAttributes);
```
Comment (karthik-tarento) [inline service/src/main/java/org/sunbird/dao/otp/impl/OTPDaoImpl.java:114]: Null check is required

**Review by karthik-tarento — CHANGES_REQUESTED** [review]

File: controller/app/controllers/otp/validator/OtpRequestValidator.java (hunk 1 above)
Comment (tarentomaheshvakkund) [reply controller/app/controllers/otp/validator/OtpRequestValidator.java]: Refractorded the methods.

File: core/platform-common/src/main/java/org/sunbird/operations/ActorOperations.java (hunk 2 above)
Comment (tarentomaheshvakkund) [reply core/platform-common/src/main/java/org/sunbird/operations/ActorOperations.java]: 1. GENERATE_OTP_V3 is used to call the generateOTPV3(request) method in the onReceive(Request request) in the OTPActor so this enum would be needed.  
2. Have updated the value of OTPCRT -> OTPCRT3.

File: service/src/main/java/org/sunbird/actor/otp/OTPActor.java (hunk 3 above)
Comment (tarentomaheshvakkund) [reply service/src/main/java/org/sunbird/actor/otp/OTPActor.java]: 1. Null check is implemented.
2. If it is a null value it is handled as empty string.

File: service/src/main/java/org/sunbird/dao/otp/impl/OTPDaoImpl.java (hunk 5 above)
Comment (tarentomaheshvakkund) [reply service/src/main/java/org/sunbird/dao/otp/impl/OTPDaoImpl.java:114]: 1. Null check is handled in the OTPActor class,if it is null empty string is recieved.

File: service/src/main/java/org/sunbird/actor/otp/OTPActor.java (hunk 4 above)
Comment (tarentomaheshvakkund) [reply service/src/main/java/org/sunbird/actor/otp/OTPActor.java:308]: 1. If contextAttributes is null then it is treated as empty string and value is saved as empty string.
2. Tested in local is working fine.

**Pull request merged**